### ISSUE TITLE
Add searx/settings.py ?

### ISSUE THREAD

**Issue by dalf** [searx/searx#2252]
Currently [`searx/__init__.py`](https://github.com/searx/searx/blob/da8b227044f45127f705f6ea94a72d368eea73bb/searx/__init__.py#L30-L51) reads `searx/settings.yml`  into `searx.settings`.

Then `searx.settings` is read directly in numerous places (not exhaustive) :
* most of https://github.com/searx/searx/blob/master/searx/__init__.py
* https://github.com/searx/searx/blob/da8b227044f45127f705f6ea94a72d368eea73bb/searx/webapp.py#L93
* https://github.com/searx/searx/blob/da8b227044f45127f705f6ea94a72d368eea73bb/searx/webapp.py#L115
* https://github.com/searx/searx/blob/da8b227044f45127f705f6ea94a72d368eea73bb/searx/preferences.py#L318-L375
* https://github.com/searx/searx/blob/da8b227044f45127f705f6ea94a72d368eea73bb/searx/utils.py#L484-L496
* https://github.com/searx/searx/blob/da8b227044f45127f705f6ea94a72d368eea73bb/searx/poolrequests.py#L70-L71
* https://github.com/searx/searx/blob/474d56c77f4c8bbd3d03f6feec0d2a13dcd41d40/searx/search.py#L38-L48

Wouldn't it be better to:
* add searx/settings.py
* clearly defined the default values here (for example a default settings written with dict and list in Python)
* write some helper to read the different values. Related to
      * https://github.com/searx/searx/pull/2251
      * https://github.com/searx/searx/pull/1278
* write a loader that check the validity of settings.yml
?

**Comment by return42** [comment]
I vote for! 

> [nearly all application setup is spread around the project and is done while importing modules](https://github.com/searx/searx/pull/2096#issuecomment-663826089)

**Comment by kvch** [comment]
I agree, it would be nice to have some mechanism to check settings.yml. I think it could be something similar to the current `preferences.py`. We could even reuse the `*Setting` classes implemented there.

If we add the ability to check the settings.yml, we could also add a subcommand or whatever to let instance admins check the configuration without starting searx.

**Comment by asciimoo** [comment]
I don't fully agree. It's true, it would make settings handling better, but I think, there are much more important tasks to solve like engine fixes or other open bugs/enhancements. It generated quite a few issues over the years, so this is a low priority task for me.

> If we add the ability to check the settings.yml, we could also add a subcommand or whatever to let instance admins check the configuration without starting searx.

Yeah, this could be a handy feature, but nobody asked for it in the past 6-7yrs, so I don't think that it would have so much added value.

> nearly all application setup is spread around the project and is done while importing modules

I still don't get why is it a problem and how would it be solved by the proposed functionality. This is a single entry point application where every module initializes itself at import time. What is wrong with that in case of searx and what would be the added value of changing this behavior?

**Comment by return42** [comment]
> and how would it be solved by the proposed functionality. 

By  ..
>> add searx/settings.py

and

>> write a loader 

> I still don't get why is it a problem

IMO it is conceptional wrong; 

a. reading or writing files just because a module was imported at some point is programming by _side effects_ (For me, it is a kind of antipattern)

b. settings.yml are the settings of the (web) application, not the settings of ever single module.  Each python module should stand for its own and the application assembles and parametrize them explicit, not by uncalled site effects (while importing).

> single entry point application where every module initializes itself at import time. 

This is what I mean in my other comment; its all backed together and not really modular ... I can't use a module stand alone, I always have to use the whole application .. and on the opposite .. I can't replace one component (e.g. how to handle the settings) in the application without touching various modules.

This might be OK for small applications, but from my POV searx becomes more and more a real "server".

**Comment by asciimoo** [comment]
> a. reading or writing files just because a module was imported at some point is programming by side effects (For me, it is a kind of antipattern)

Why? Print or socket write is also a side effect but both are very useful. From this point of view any kind of module initialization (like defining functions/classes/variables in a module) is also a side effect, it would be hard to avoid it. Also, I don't understand, that why is it a problem in case of searx that initialization happens in import time? Why would be better to create an `init(*args)` function in every module and call it manually after importing it? What is the disadvantage of the current behavior?

> b. settings.yml are the settings of the (web) application, not the settings of ever single module. Each python module should stand for its own and the application assembles and parametrize them explicit, not by uncalled site effects (while importing).

Settings is an importable module which is used by multiple other modules. How would you handle it? Would you write a config file for each module?

> This is what I mean in my other comment; its all backed together and not really modular ... I can't use a module stand alone, I always have to use the whole application

Yes, and it is perfectly fine, searx is an application it isn't a framework or a lib,  modules are internal and there is no intent to use them outside of the application.

> I can't replace one component (e.g. how to handle the settings) in the application without touching various modules.

That's how dependency work, there are multiple modules depend on the settings module, just like there are multiple modules that depend on built-in modules or other external ones. If you want to replace a module that is used by an other module as a dependency, then you have to modify the importing module too, but this is true for every single dependency of every single module.
This situation wouldn't be different if we had a settings.py.

Probably it is my fault, but I still can't see the added value.

**Comment by kvch** [comment]
> I don't fully agree. It's true, it would make settings handling better, but I think, there are much more important tasks to solve like engine fixes or other open bugs/enhancements. It generated quite a few issues over the years, so this is a low priority task for me.

It is true, engine fixes should be the most important things on our TODO lists. I mean everyone can write a basic meta search engine from scratch. The real value of searx is the 70-80 engines which has been developed over the years. We should focus on that, because this sets us apart from other search engines.

>> If we add the ability to check the settings.yml, we could also add a subcommand or whatever to let instance admins check the configuration without starting searx.
>
> Yeah, this could be a handy feature, but nobody asked for it in the past 6-7yrs, so I don't think that it would have so much added value.

It is indeed not an urgent/important feature.

>>  This is what I mean in my other comment; its all backed together and not really modular ... I can't use a module stand alone, I always have to use the whole application

I am not sure why this is a problem. Searx is a standalone web application. Why would someone want to use it as a library or framework or whatever? These use cases are not supported.

**Comment by return42** [comment]
> Also, I don't understand, that why is it a problem in case of searx that initialization happens in import time?

See https://github.com/searx/searx/pull/2256#issuecomment-706672264

**Comment by return42** [comment]
> Probably it is my fault, but I still can't see the added value.

Not a fault, we have different views on this subject for practical reasons

Even if we think there is only one single app today, we already have more. Building the docs is another _application_ (better called "use case") and we do not know what our need is tomorrow. The future might introduce complete new use cases,  we can't think of today. So it is good to decouple tasks from each other. (importing is one task, initializing is another task).

In the following I will try to explain how I would decouple import from initialize ..

> Settings is an importable module which is used by multiple other modules. How would you handle it? Would you write a config file for each module?

No.  From my view, settings.yml is the settings file of the web-application.  These settings should be loaded when the Flask App is initialized just before the app is started. With the current implementation,  the  `searx.webapp.run` function might be good place for such a task:

https://github.com/searx/searx/blob/e78bfd4d8b62a521e08029bb0e708b2bcec9958d/searx/webapp.py#L1035-L1043

> Why would be better to create an init(*args) function in every module and call it manually after importing it?

The (new) settings.py should contain a loader function for tasks like this

https://github.com/searx/searx/blob/e78bfd4d8b62a521e08029bb0e708b2bcec9958d/searx/__init__.py#L50-L51

And other modules like autocomplete (to take just one simple example) should have a _init_ method.  Since there is no need for OOP in autocomplete and to keep things simple, this init is done by a function, which inits a global variable. To illustrate what I mean:

````python
settings=None

def get(*args, **kwargs):
    if 'timeout' not in kwargs:
        kwargs['timeout'] = settings['outgoing']['request_timeout']
    return http_get(*args, **kwargs)

def init(app_settings):
    global settings
    settings = app_settings
````

**Comment by dalf** [comment]
related to the issue, `static_path` is defined two different locations:
* https://github.com/searx/searx/blob/f310305c54c3cd1d9fc74f09453294edbd2b5486/searx/__init__.py#L26
  Here the path is absolute
* https://github.com/searx/searx/blob/f310305c54c3cd1d9fc74f09453294edbd2b5486/searx/webapp.py#L95
  https://github.com/searx/searx/blob/f310305c54c3cd1d9fc74f09453294edbd2b5486/searx/webutils.py#L50-L55
  Here the path may be relative

The only usage of `searx.static_path` :
https://github.com/searx/searx/blob/f310305c54c3cd1d9fc74f09453294edbd2b5486/searx/plugins/__init__.py#L127

----

This value is never used:
https://github.com/searx/searx/blob/f310305c54c3cd1d9fc74f09453294edbd2b5486/searx/__init__.py#L25

-----

https://github.com/searx/searx/blob/f310305c54c3cd1d9fc74f09453294edbd2b5486/searx/webapp.py#L818

`settings['server']['base_url']` should be `get_base_url()`  ?

**Comment by dalf** [comment]
For reference, see 
https://github.com/dalf/searx/tree/add_settings_default
more precisely this commit: https://github.com/dalf/searx/commit/23b6659a0d53061132ceb997746740c4cc6701e3

**Comment by return42** [comment]
> more precisely this commit: dalf@23b6659

I cloned your add_settings_default branch and made a first review and some small tests .. looks good to me, what I really like is the cleanup of the `static_path` mess .. do you like to rebase and send PR?
